pkg/store/storepb/custom.go

```diff
@@ -456,3 +475,138 @@ func CompareLabels(a, b []Label) int {
 func LabelsToPromLabelsUnsafe(lset []Label) labels.Labels {
 	return labelpb.ZLabelsToPromLabels(lset)
 }
+
+// Type alias because protoc-go-inject-field does not support
+// managing imports.
+type syncPool = sync.Pool
+
+// Close returns the memory used for marshaling, if any.
+func (m *SeriesResponse) Close() {
+	if m == nil || m.respBuf == nil {
+		return
+	}
+
+	m.respPool.Put(m.respBuf)
+	m.respBuf = nil
+}
+
+// The following were copied/pasted from gogoprotobuf generated code with changes
+// to make it work with sync.Pool / []byte slice.
+func (m *SeriesResponse) Marshal() (dAtA []byte, err error) {
+	size := m.Size()
+
+	var respBuf []byte
+
+	// No pool defined, allocate directly.
+	if m.respPool == nil {
+		respBuf = make([]byte, size)
+	} else {
+		if m.respBuf == nil {
+			poolBuf := m.respPool.Get()
+			if poolBuf == nil {
+				respBuf = make([]byte, size)
+				m.respBuf = &respBuf
+			} else {
+				m.respBuf = poolBuf.(*[]byte)
+				respBuf = *m.respBuf
+			}
+		} else {
+			if cap(*m.respBuf) < size {
+				if m.respPool != nil {
```

Comment:
Not sure if we have to put too small one back -if it was too small, most likely it will be too small for other responses too.

Not a blocker, it should work both ways, just guessing here a bit.